sql-serializer/src/main/java/io/cdap/pipeline/sql/serializer/bigquery/BigQuerySQLSerializer.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.pipeline.sql.serializer.bigquery;
+
+import io.cdap.pipeline.sql.api.Column;
+import io.cdap.pipeline.sql.api.Filter;
+import io.cdap.pipeline.sql.api.StructuredQuery;
+import io.cdap.pipeline.sql.api.Table;
+import io.cdap.pipeline.sql.api.constants.DateTimeConstant;
+import io.cdap.pipeline.sql.api.constants.IntegerConstant;
+import io.cdap.pipeline.sql.api.constants.StringConstant;
+import io.cdap.pipeline.sql.api.enums.OperandType;
+import io.cdap.pipeline.sql.api.enums.PredicateOperatorType;
+import io.cdap.pipeline.sql.api.enums.QueryableType;
+import io.cdap.pipeline.sql.api.interfaces.Constant;
+import io.cdap.pipeline.sql.api.interfaces.Operand;
+import io.cdap.pipeline.sql.api.interfaces.Queryable;
+import io.cdap.pipeline.sql.serializer.interfaces.SQLSerializer;
+
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Serializes a {@link StructuredQuery} into a {@link String} able to be executed upon Google BigQuery.
+ */
+public class BigQuerySQLSerializer implements SQLSerializer {
+  public String getSQL(StructuredQuery query) {
+    if (query == null) {
+      throw new IllegalArgumentException("Query cannot be null");
+    }
+    StringBuilder b = new StringBuilder();
+    return appendQuerySQL(b, query, false).toString();
+  }
+
+  public String getSQL(StructuredQuery query, Table temporaryTable) {
+    if (query == null) {
+      throw new IllegalArgumentException("Query cannot be null");
+    }
+    if (temporaryTable.hasAlias()) {
+      throw new IllegalArgumentException("Cannot have alias on temporary table");
+    }
+    if (temporaryTable.hasProject() || temporaryTable.hasDatabase()) {
+      throw new IllegalArgumentException("Temporary tables cannot be fully qualified");
+    }
+    StringBuilder b = new StringBuilder();
+    b.append("CREATE TEMPORARY TABLE ");
+    appendTableSQL(b, temporaryTable, false);
+    b.append(" AS ");
+    appendQuerySQL(b, query, false);
+    return b.toString();
+  }
+
+  private StringBuilder appendQuerySQL(StringBuilder b, StructuredQuery query, boolean isSubquery) {
+    // Keep track of which columns were defined in the current nesting level since they may also be used in filter
+    HashSet<Column> definedColumns = new HashSet<>();
```

Comment:
I also don't really understand what this is used for. From the comment it sounds like if the filter calls it one thing but the select has it aliased to something else? If so, I don't think that's valid sql. For example:
```
select name as first_name from customers where first_name is not null;
```
is not a valid query (at least in MySQL). The where clause needs to refer to 'name'.

Reply:
Yes, you are correct, I should have provided an example to explain the purpose of this. Since the query serialization is recursive, the query thus contains a set of aliased columns which are passed down to the column serialization functions. Since, as per your example, an alias may not be used when using a column at the query level, we keep track of the columns which are aliased to prevent the column serializer from converting the column to its alias form.

Now that I think of it, this may be unnecessary because, assuming we declare separate columns for all query levels, a column should never be serialized to its alias form. In that case, this logic would be unnecessary.

Reply:
Fixed by removing the definedColumns set